### PR TITLE
[C][Client] Free list or map memory when json parsing fails

### PR DESCRIPTION
**Pull request by ityuhui**
When a list or map JSON cannot to be parsed successfully,  the original code will goto the end of function `*_parseFromJSON` without freeing the memory of the list or map.

This PR fixes this memory leak.

@wing328 @zhemant @michelealbano

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/C-libcurl/model-body.mustache [hunk 1]
```diff
@@ -904,6 +920,48 @@ end:
...
+    {{/isPrimitiveType}}
+    {{/isArray}}
+    {{#isMap}}
+    if ({{{name}}}List) {
+        listEntry_t *listEntry = NULL;
+        list_ForEach(listEntry, {{{name}}}List) {
+            keyValuePair_t *localKeyValue = (keyValuePair_t*) listEntry->data;
+            free(localKeyValue->key);
+            localKeyValue->key = NULL;
+            free(localKeyValue->value);
```
[the remark this answers is not in the document]
Comment (ityuhui) [reply modules/openapi-generator/src/main/resources/C-libcurl/model-body.mustache:955]: Thanks. I'll deal by data type.

**Comment by minerba** [comment]
@ityuhui 
Thanks your PR. Is it necessary to Fix some Data Type? regarding the Map

File: modules/openapi-generator/src/main/resources/C-libcurl/model-body.mustache (hunk 1 above)
[the remark this answers is not in the document]
Comment (ityuhui) [reply modules/openapi-generator/src/main/resources/C-libcurl/model-body.mustache:972]: Done. Can you take another look ?

**Comment by ityuhui** [comment]
Hi @zhemant 

Can you check the new code changes? I have updated based on your comments.

**Comment by zhemant** [comment]
Thank you for the changes @ityuhui. 

If we are planning to support primitive types(string, byteArray, enum, number, boolean) then these changes are sufficient, I would recommend we add a non-primitive section where we add a comment line `data type not supported` as a note that we do not support it in C generator currently.

https://swagger.io/docs/specification/data-models/dictionaries/
`Values can be primitives (strings, numbers or boolean values), arrays or objects`

**Comment by ityuhui** [comment]
> I would recommend we add a non-primitive section where we add a comment line `data type not supported` as a note that we do not support it in C generator currently.
> 
> https://swagger.io/docs/specification/data-models/dictionaries/ `Values can be primitives (strings, numbers or boolean values), arrays or objects`

Thanks @zhemant 

I address this comment by the commit https://github.com/OpenAPITools/openapi-generator/pull/11866/commits/a76b542d5015d8a7eabf0bc541a9dd09124d7b8d 
PTAL

**Comment by ityuhui** [comment]
I have logged a bug https://github.com/kubernetes-client/c/issues/112 for the non-primitive in map.

I will fix it later.

**Comment by ityuhui** [comment]
Is this PR ready to merge now ? @zhemant

**Comment by zhemant** [comment]
Thanks for the changes @ityuhui . LGTM

**Review by zhemant — APPROVED** [review]

**Comment by wing328** [comment]
Tested locally and C petstore built without issues.

**Comment by ityuhui** [comment]
Thanks @wing328 @zhemant

**Pull request merged**